### DISCUSSION TITLE
Uploading data to NCBI SRA

### DISCUSSION THREAD

**Discussion by yaaminiv** [RobertsLab/resources#1400] [Q&A] [answered]
@kubu4  How do you usually upload raw data to NCBI SRA? I'm trying to upload the *C. gigas* Manchester files through the Aspera plug-in and it keeps timing out.

**Comment by kubu4** [comment, marked as answer]
I always use FTP.

**Reply by yaaminiv** [reply]
Through the command line? I know more recent Mac OS versions don't have FTP command line options anymore.

**Reply by kubu4** [reply]
> Through the command line?

Yep.

> I know more recent Mac OS versions don't have FTP command line options anymore.

Via Homebrew, you can install FTP (and other tools) like so:

`brew install inetutils`

**Reply by yaaminiv** [reply]
> Via Homebrew, you can install FTP (and other tools)

I always forget about Homebrew! Thanks for the tip.

When using FTP, are the files you're transferring on your local machine or on `nightingales`? I'm assuming having owl mounted on my computer and going from there would make the process slower?

**Reply by kubu4** [reply]
Usually Owl mounted on my computer (almost always a computer that's in the lab, on the same local network as Owl). Owl doesn't have a native `ftp` command line program installed (and, as far as I can tell, there isn't one that can be easily installed), so you sort of to do it this way.

Regardless, FTP transfers are SLOW, even if you have the files locally. I don't know how many files you have, but you should plan on the transfer taking all day, if not longer.

Pro Tip: When logging into NCBI servers, add the `-i` argument to the ftp command. This will then allow you to use `mput` when you want to put _multiple_ files (thus, the `m` in `mput`) with a command like: `mput *.gz`. Otherwise, you have to do them one by one using the `put` command over and over again.

**Reply by yaaminiv** [reply]
> Pro Tip: When logging into NCBI servers, add the -i argument to the ftp command. 

Will do!

> Regardless, FTP transfers are SLOW, even if you have the files locally. I don't know how many files you have, but you should plan on the transfer taking all day, if not longer.

Yikes, I have 16 files and I'm doing this from my laptop since we don't have any lab computers...I assume this will take multiple days?

**Reply by kubu4** [reply]
Probably a day or two, yes. I also frequently encounter dropped connection issues, which will kill the transfer. FTP _will not_ resume from a previous transfer, so if the connection gets dropped, you'll have to take note of which file was currently transferring and transfer that file again (as well as any remaining). 

It's an annoying process.

**Reply by kubu4** [reply]
Since you'll likely want to use your laptop and take it places with you, I'd initiate the FTP transfer from one of the Roberts Lab computers using `screen`. `screen` allows you to kill a remote connection, but the processes running on the computer keep going. So, the process would be:

1. SSH into a Roberts Lab computer.
2. Mount Owl (if you haven't already) somewhere.
3. Start `screen`: `screen`
4. Navigate to desired directory on Owl.
5. Initiate FTP to NCBI.

Then, you can exit the connection to the Roberts Lab computer any time and SSH back in and check in on your `screen` session at any time.

**Reply by yaaminiv** [reply]
> Mount Owl (if you haven't already) somewhere.

Can I have owl mounted on my personal laptop, or does it need to be on a Roberts Lab computer?

**Reply by kubu4** [reply]
I think you can mount it on your laptop, but you won't be able to shut your laptop down or disconnect it from the network during the FTP transfer. That's why I recommended SSH-ing to a Roberts Lab computer to do this.

**Reply by yaaminiv** [reply]
Got it! And how do I mount owl on a Roberts Lab computer I'm SSH-ing into...

**Reply by kubu4** [reply]
Use this example:

`sudo mount --verbose -t cifs //owl.fish.washington.edu/web /mnt/owl/ -o credentials=/home/<user>/.owl_creds,vers=1.0`

Here's what's happening:

- `sudo mount --verbose`: Telling computer to mount a drive in verbose mode. You may not need to use `sudo` - see below.

- `-t cifs`: File system mount type.

- `//owl.fish.washington.edu/web/nightingales`: Location on Owl that you want to mount. 

- `/mnt/owl/`: Location of empty folder on the computer you're logged into. Preferably, you should create a dedicated directory in your `/home/` directory. That should allow you to run the `mount` command without `sudo`.

- `-o credentials=/home/<user>/.owl_creds`: Path to a "hidden" credentials file. Contains username and password info for your Owl login. Formatted like this:

    ```shell
       username=<username>
       password=<password>
     ```

    IMPORTANT: This method isn't very secure. If you use this method, be sure to change the permissions on that credentials file:

    - `chmod 700 .owl_creds`

- `,vers=1.0`: Specifies SMB version to use, since Owl is using an older version.

**Comment by yaaminiv** [comment]
> /mnt/owl/: Location of empty folder on the computer you're logged into. Preferably, you should create a dedicated directory in your /home/ directory. That should allow you to run the mount command without sudo.

So I create a folder (ex. `owl`) in my `home` directory, then change `/mnt/owl/` to that path?

> This method isn't very secure. If you use this method, be sure to change the permissions on that credentials file

Is there a way to log in without using this method?

**Reply by kubu4** [reply]
> So I create a folder (ex. owl) in my home directory, then change /mnt/owl/ to that path?

Yep.

> This method isn't very secure.

To clarify, it's secure in that credentials are encrypted when being sent over the internet to the server and all that. It's _not_ secure because login credentials are stored in a non-encrypted, plain text file on your computer. So, if someone was to log into your laptop, view that file, _and_ correctly deduce that it was to be used for `owl.fish.washington.edu`, then they would have access to your user account on Owl. 

So, by me saying "it isn't very secure" is more of a "cybersecurity" good practices violation.

However, if you don't want to create/use a credentials file, then you'd modify the command to look like this:

`sudo mount --verbose -t cifs //owl.fish.washington.edu/web /mnt/owl/ -o username=<owl_username>,vers=1.0`

I believe it will prompt you for your password after that.

**Reply by yaaminiv** [reply]
I'm SSH-ed into `ostrich` and tried `mount --verbose -t cifs //owl.fish.washington.edu/web /owl/ -o username=<yaaminiv>,vers=1.0`:

> -bash: yaaminiv: No such file or directory

Similar error with `sudo mount --verbose -t cifs //owl.fish.washington.edu/web /owl/ -o credentials=/home/<user>/.owl_creds,vers=1.0`

> -bash: user: No such file or directory

**Reply by yaaminiv** [reply]
WHOOPS hang on did this: 

`sudo mount --verbose -t cifs //owl.fish.washington.edu/web /owl/ -o credentials=/home/yaaminiv/.owl_creds,vers=1.0`

and now have the following error: srlab is not in the sudoers file.  This incident will be reported.

**Reply by yaaminiv** [reply]
Tried running without `sudo`:

`mount --verbose -t cifs //owl.fish.washington.edu/web owl/ -o credentials=/home/yaaminiv/.owl_creds,vers=1.0`

Got this error: 

```
mount: illegal option -- -
usage: mount [-dfruvw] [-o options] [-t external_type] special node
       mount [-adfruvw] [-t external_type]
       mount [-dfruvw] special | node
```

**Reply by kubu4** [reply]
My instructions were for Linux, not a Mac; sorry about that. Not entirely sure what to do here. However, the error is related to the use of `--verbose`. Get rid of that and see what happens...

**Reply by yaaminiv** [reply]
```
ostrich:~ srlab$ mount -t cifs //owl.fish.washington.edu/web owl/ -o credentials=/home/yaaminiv/.owl_creds,vers=1.0
usage: mount [-dfruvw] [-o options] [-t external_type] special node
       mount [-adfruvw] [-t external_type]
       mount [-dfruvw] special | node
ostrich:~ srlab$ sudo mount -t cifs //owl.fish.washington.edu/web /owl/ -o credentials=/home/yaaminiv/.owl_creds,vers=1.0
Password:
srlab is not in the sudoers file.  This incident will be reported.
```

> My instructions were for Linux

Tried SSH into roadrunner:

`sudo mount -t cifs //owl.fish.washington.edu/web owl/ -o username=yaaminiv,vers=1.0`

```
srlab@roadrunner:~$ cd owl
srlab@roadrunner:~/owl$ ls
Athaliana    index.html                        phainopepla   trochophora
dlf          index.php                         porites       web_images
eimd         kaitlyn                           sam           wetgenes
fish546      lilmax                            scaphapoda    whale
generosa     nightingales                      spartina
halfshell    O_lurida_genome_assemblies_BGI    symbiodinium
hesperornis  P_generosa_genome_assemblies_BGI  testing
```

SUCCESS (so far). I'll try using `screen` now that I have `owl` mounted!

**Comment by yaaminiv** [comment]
Once I mounted `owl` on `roadrunner`, I was able to use `screen` and `ftp -i`!

I exited my terminal window and logged back in, but I can't find my `screen` session...I logged back into `roadrunner` via SSH, navigated to my owl directory, and started `screen`, but all I see is a blank Terminal window. How do I check on my session?

**Reply by kubu4** [reply]
Exit the current screen session:

`exit`

Should print out something like:

```shell
[screen is terminating]
```

Then, list active screen sessions:

```shell
screen -ls
```

Attach the screen session you want:

```shell
screen -r <number>
```

The number will be shown at the beginning of `.pts-11.roadrunner (02/15/2022 01:50:14 PM)        (Attached)` when you run the `screen -ls` command.

**Reply by sr320** [reply]
omg this is exciting 🥰 - will she get in or not? This could definitely be a movie trailer 😆 I feel like there is a bomb countdown ...

**Reply by kubu4** [reply]
LOL!

**Reply by yaaminiv** [reply]
Thanks for the `screen` tutorial (that I totally could have googled myself whoops)! FTP is still going.
